text/3498-lifetime-capture-rules-2024.md

```diff
@@ -0,0 +1,642 @@
+
+In the trait impls, AFIT will automatically capture all type and lifetime parameters from the outer impl and from the associated function or method signature.  This ensures that signatures are copyable from trait definitions to impls.
+
+This behavior of AFIT will be parsimonious with the current stable capture behavior of `async fn` in inherent impls.
+
+For example:
+
+```rust
+#![feature(async_fn_in_trait)]
+
+trait Trait<'t>: Sized {
+    async fn foo<'f>(self, x: &'t (), y: &'f ()) -> (Self, &'t (), &'f ());
+    //                                              ^^^^^^^^^^^^^^^^^^^^^^
+    // Method signature lifetimes, trait input lifetimes, and
+    // lifetimes in the Self type may all be captured in this opaque
+    // type in the impl.
+}
+
+struct Foo<'s>(&'s ());
+impl<'t, 's> Trait<'t> for Foo<'s> {
+    async fn foo<'f>(self, x: &'t (), y: &'f ()) -> (Foo<'s>, &'t (), &'f ()) {
+        //                                          ^^^^^^^^^^^^^^^^^^^^^^^^^
+        // The opaque type captures:
+        //
+        //   - `'f` from the method signature.
+        //   - `'t` from the outer impl and a trait input lifetime.
+        //   - `'s` from the outer impl and the Self type.
+        (self, x, y)
+    }
+}
+```
+
+This updates and supersedes the behavior specified in [RFC 3425].
+
+# Acknowledgments
+
+Thanks to Tyler Mandry (@tmandry) for his collaboration on the earlier design document for the 2024 lifetime capture rules, and thanks to Michael Goulet (@compiler-errors) for helpful discussions and insights on this topic.
+
+All errors and omissions remain those of the author alone.
+
+# Appendix A: Other resources
+
+Other resources:
+
+- [Lifetime capture rules 2024 T-lang design meeting](https://hackmd.io/sFaSIMJOQcuwCdnUvCxtuQ)
+- [Capturing lifetimes in RPITIT](https://hackmd.io/zgairrYRSACgTeZHP1x0Zg)
+
+# Appendix B: Matrix of capturing effects
+
+| | 2021: *Outer LP* | 2021: *Item LP* | 2024: *Outer LP* | 2024: *Item LP* |
+|-|-|-|-|-|
+| RPIT          | N   | N | Y   | Y |
+| `async fn`    | Y   | Y | Y   | Y |
+| GATs          | Y   | Y | Y   | Y |
+| TAIT          | N/A | Y | N/A | Y |
+| ATPIT         | Y   | Y | Y   | Y |
+| RPITIT: trait | Y   | Y | Y   | Y |
+| RPITIT: impl  | Y   | Y | Y   | Y |
+
+In the table above, "LP" refers to "lifetime parameters".
+
+The 2024 behavior described for all items is the behavior under this RFC.
+
+The 2021 behavior described for RPIT and `async fn` is the stable behavior in Rust 2021.  The other 2021 behaviors described are the behaviors that will be implemented for the features ahead of stabilization.
+
+*All* of the features above automatically capture all lifetimes from all type parameters in scope in both the 2021 and the 2024 editions.
+
+# Appendix C: The 2021 edition rules fail for RPITIT
```

Comment:
I don't see why this RFC should deviate so much from the RFC template. This and other sections could go under Rationale and Alternatives, while the Future possibilities could go under the corresponding heading in the template.

Reply:
This is a big document, and two of the goals I had in mind were to keep the changes from the design document contained and easy to review and to get the document posted on a timeline relevant for the RPITIT stabilization.  I'll have another look at this, but I am still concerned about growing the diff too much from the design document.

Reply:
Hmm, I don't know of anyone who's doing a literal diff between that doc and this one (though maybe they would, if it were convenient and useful to show not much has changed). In terms of the changes needed I think it's just rearranging the existing sections. Which would indeed make the diff larger.

Reply:
There is the mechanical diff, but I had in mind more the mental diff.  It seems valuable for this to look like the same document and for the major sections and appendices to mostly line up.  As mentioned, I'll review it.  There is some flow between the various sections and appendices, and even aside from trying to keep the document structurally similar to aid familiarity, it's easy to make such a document worse by moving things around, so I just want to be careful about that.  It seems more important that the document is maximally clear and readable than that it maximally follows the template.  It's not the only RFC that has diverged here.  But I'll see about trying to close the distance further.

Reply:
If you don't mind, I'm going to mark this as resolved.  I took a pass at doing this.  However, given the existing and ongoing conversations around this document, I want to preserve the liveness of inbound links to the various appendices.  This requires keeping stub appendices with forward links to the new sections after moving the content.  Combined with the fact that there are many appendices that remain after the rearrangement, the result didn't feel like a net improvement.  I'll keep this in mind, though, in case I think of a better approach.